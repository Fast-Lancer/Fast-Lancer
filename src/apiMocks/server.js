import { rest } from 'msw'
import { setupServer } from 'msw/node'

const url = process.env.REACT_APP_SUPABASE_URL + '/rest/v1'

let postedClient = {}
let postedProject = {}

export const server = setupServer(
  rest.get(url + '/clients', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(select === '*' && !id) {
      // response for getClients()
      return res(
        ctx.json([
          {
            id: 1,
            client_name: 'bob1',
            business_name: 'business1'
          },
          {
            id: 2,
            client_name: 'bob2',
            business_name: 'business2'
          },
          {
            id: 3,
            client_name: 'bob3',
            business_name: 'business3'
          }
        ])
      )
    } else if(id) {
      // response for getClient(id)
      if(id === 'eq.42') {
        // Janky workaround to make testing edit/create possible.
        // The post route mock will set the id to 42
        // So this responds with the posted data when client.id is 42
        return res(
          ctx.json(
            postedClient
          )
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              client_name: 'bob1',
              business_name: 'business1'
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/clients', (req, res, ctx) => {
    let [client] = req.body
    client = { ...client, id: 42 }
    postedClient = client
    return res(
      ctx.json([
        client
      ])
    )
  }),
  rest.get(url + '/projects', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(!id) {
      // response for getProjects()
      return res(
        ctx.json([
          {
            id: 1,
            title: 'project1',
            client_id: '1',
            clients: {
              client_name: 'bob1'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          },
          {
            id: 2,
            title: 'project2',
            client_id: '2',
            clients: {
              client_name: 'bob2'
            },
            
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          
          },
          {
            id: 3,
            title: 'project3',
            client_id: '3',
            clients: {
              client_name: 'bob3'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          }
        ])
      )
    } else if(id) {
      // response for getProjectById(id)
      if(id === 'eq.42') {
        return res(
          ctx.json(postedProject)
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              title: 'project1',
              client_id: '1',
              clients: {
                client_name: 'bob1'
              },
              date_start:'2022-01-20',
              date_end:'2022-01-21',
              URL: 'something@something.com',
              description: 'ok',
              notes: 'test notes',
              hourly_rate: 33,
              hours_quoted: 22,
              price_quoted: 300
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/projects', (req, res, ctx) => {
    let project = req.body
    project = { ...project, id: 42 }
    postedProject = project
    return res(
      ctx.json([project])
    )
  })
)
