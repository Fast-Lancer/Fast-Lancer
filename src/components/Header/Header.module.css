.header {
  display: grid;
  grid-template: 1fr 1fr 1fr / 2fr 8fr 2fr;
  height: 150px;
}

.header img {
  height: 80%;
  width: 80%;
}

.header > nav {
  background-color: rgb(180, 180, 180);
  border: 1px solid rgb(82, 81, 81);
  margin: 10px;
  border-radius: 10px;
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  align-items: center;
  gap: 10px;
  font-size: 18px;
}

.logoDiv {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  grid-row: 1 / 4;
}

.hamburger {
  display: none;
  margin: 5px;
  grid-row: 1 / 2;
  grid-column: 3 / 4;
  width: 40px;
  height: 40px;
}

.bar {
  display: block;
  width: 25px;
  height: 3px;
  margin: 5px auto;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  background-color: #101010;
}

.bar {
  display: block;
  width: 29px;
  height: 3px;
  margin-bottom: 5px;
  position: relative;
  background: black;
  border-radius: 3px;
  z-index: 1;
  transform-origin: 4px 0px;
  transition: transform 0.5s cubic-bezier(0.77,0.2,0.05,1.0),
              background 0.5s cubic-bezier(0.77,0.2,0.05,1.0),
              opacity 0.55s ease;
}

.hamburger .bar:first-child {
  transform-origin: 0% 1px;
}

.hamburger .bar:nth-last-child(1) {
  transform-origin: 0% 100%;
}

.bar.active {
  opacity: 1;
  transform: translate(3px, -3px) scale(1.15, 1.15) rotate(45deg);
  background: #232323;
}

.hamburger .bar.active:nth-last-child(2) {
  opacity: 0;
  transform: rotate(0deg) scale(0.2, 0.2);
}

.hamburger .bar.active:nth-last-child(1) {
  transform: translate(4px, 4px) scale(1.15, 1.15) rotate(-45deg);
}

@media(max-width: 768px) {
  .header {
    height: 50px;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
  }

  .header > nav {
    position: fixed;
    top: -100%;
    flex-direction: column;
    background-color: #fff;
    width: 100%;
    border-radius: 0px;
    text-align: center;
    transition: all .3s ease-in-out;
    box-shadow: 0 10px 27px rgba(0, 0, 0, 0.05);
    background-color: grey;
    margin: 0px;
  }

  .header > nav.active {
    top: 0;
  }

  .logoDiv {
    width: 50px;
    height: 50px;
  }
  
  .hamburger {
    display: block;
    z-index: 1;
  }

  .header img {
    width: 50px;
    height: 50px;
  }
}
