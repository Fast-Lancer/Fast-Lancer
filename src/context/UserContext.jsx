import { createContext, useContext, useMemo, useState } from 'react'
import { getUser } from '../services/users'

const UserContext = createContext({})

export function UserProvider({ children }) {
  const currentUser = getUser()
  const [user, setUser] = useState(currentUser ?? {})
  const memo = useMemo(() => ({ user, setUser }), [user])
  return (
    <UserContext.Provider value={memo}>
      {children}
    </UserContext.Provider>
  )
}

export function useUser() {
  const context = useContext(UserContext)

  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider')
  }

  return context
};
