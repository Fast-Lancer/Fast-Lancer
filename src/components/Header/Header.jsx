import { Link, useLocation } from "react-router-dom";
import FastLancerLogo from "../../assets/FastLancerLogo.svg";
import { useUser } from "../../context/UserContext.jsx";

export default function Header() {
  const { user, setUser } = useUser();
  const location = useLocation();
  const currentPage = location.pathname;

  return (
    <div>
      <img src={FastLancerLogo} alt="fast lancer logo" />
      <div>
        {currentPage === "/login" || currentPage === "/signup" ? (
          <>
            <Link to="/login">Log In</Link>
            <Link to="/signup">Sign Up</Link>
          </>
        ) : (
          <>
            <Link to="/clients">Clients</Link>
            <Link to="/projects">Projects</Link>
          </>
        )}
        {user?.id ?? (
          <Link onClick={() => setUser({})} to="/login">
            Log Out
          </Link>
        )}
      </div>
    </div>
  );
}
