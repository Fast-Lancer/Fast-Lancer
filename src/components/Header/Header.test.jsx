import { useEffect } from 'react'
import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { UserProvider, useUser } from '../../context/UserContext.jsx'
import { server } from '../../apiMocks/server.js'
import Header from './Header'

beforeAll(() => {
  server.listen()
})

afterAll(() => {
  server.close()
})

const UserAdder = () => {
  const { setUser } = useUser()
  useEffect(() => {
    setUser({ id: 1 })
  }, [])

  return null
}

it('renders the Header component accurately', () => {
  const { container } = render(
    <MemoryRouter initialEntries={['/']}>
      <UserProvider>
        <UserAdder>
          <Header />
        </UserAdder>
      </UserProvider>
    </MemoryRouter>

  )
  expect(container).toMatchSnapshot()
})
