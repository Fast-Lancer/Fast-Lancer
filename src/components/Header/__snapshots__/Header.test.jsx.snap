// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Header component accurately 1`] = `<div />`;
